docs/couler-api-design.md

```diff
@@ -0,0 +1,87 @@
+# Couler API Design
+
+This document outlines the design of core Couler APIs to support multiple workflow backends.
+
+## Goals
+
+* Design the core Couler APIs that can be implemented across different workflow engines.
+* Provide a *minimal* set of APIs that are forward-looking, engine-agnostic, and less likely to deprecate over-time.
+
+## Non-Goals
+
+* Provide an exhaustive set of APIs that cover all use cases and all workflow engines.
+* Provide implementation details.
+
+## Design
+
+Core operations (`couler.ops`):
+
+* `run_step(step_def)` where `step_def` is the step definition that can be a container spec, Python function,
+or spec that's specific to the underlying workflow engine (e.g. k8s CRD if Argo Workflow is used).
+
+Control flow (`couler.control_flows`):
+
+* `map(func, *args, **kwargs)` where `*args` and `**kwargs` correspond to various arguments passed to the function `func`.
+* `when(cond, if_op, else_op)` where
+    * `cond` can be any of the following predicates: `equal()`, `not_equal()`, `bigger()`, `smaller()`, `bigger_equal()`, `smaller_equal()`.
+    * The operation defined in `if_op` will be executed when `cond` is true. Otherwise, `else_op` will be executed.
+* `while_loop(cond, func, *args, **kwargs)` where
```

Comment:
naming suggestion: while_loop -> loop?

Reply:
"loop" by itself might be a bit confusing as it can be interpreted as "for loop"